File: Nechifor Georgian/Day 33/Easy.java
class ListNode {
    int value;
    ListNode next;
    ListNode(int x) {
        value = x;
    }
}

public class Easy {
    /*
     * DAY 2
     *
     */

    //https://leetcode.com/problems/merge-two-sorted-lists/
    private ListNode insertLast(ListNode node, int x) {
        ListNode aux = node;
        if (node == null)
            return new ListNode(x);
        while (aux.next != null) {
            aux = aux.next;
        }
        aux.next = new ListNode(x);
        return node;
    }

    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        ListNode rez = null;
        while (l1 != null && l2 != null) {
            if (l1.val < l2.val) {
                rez = insertLast(rez, l1.val);
                l1 = l1.next;
            } else {
                rez = insertLast(rez, l2.val);
                l2 = l2.next;
            }
        }

        while (l1 != null) {
            rez = insertLast(rez, l1.val);
            l1 = l1.next;
        }

        while (l2 != null) {
            rez = insertLast(rez, l2.val);
            l2 = l2.next;
        }

        return rez;
    }

    //https://leetcode.com/problems/linked-list-cycle/
    public boolean hasCycle(ListNode head) {
        if (head == null || head.next == null) return false;
        while (head != null) {
            if (head.val != Integer.MIN_VALUE)
                head.val = Integer.MIN_VALUE;
            else
                return true;

            head = head.next;
        }
        return false;
    }

    //https://leetcode.com/problems/min-stack/
    class MinStack {
        private static final int MAX = 1000;
        private int top;
        private int[] a = new int[MAX];

        MinStack() {
            top = -1;
        }

        public boolean isEmpty() {
            return top < 0;
        }

        //insert last
        public void push(int x) {
            if (top >= MAX - 1)
                return;
            else {
                a[++top] = x;
            }
        }

        public void pop() {
            if (top < 0)
                return;
            else {
                int x = a[top--];
            }

        }

        public int top() {
            if (top < 0) return -1;
            return a[top];
        }

        public int getMin() {
            if (top < 0) return 0;
            int minValue = Integer.MAX_VALUE;
            for (int i = 0; i <= top; i++)
                minValue = Integer.min(minValue, a[i]);
            return minValue;
        }
    }

    //https://leetcode.com/problems/majority-element/
    public int majorityElement(int[] nums) {
        int majority = 0;
        Map<Integer, Integer> freq = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            if (freq.containsKey(nums[i])) {
                freq.put(nums[i], freq.get(nums[i]) + 1);
            } else
                freq.put(nums[i], 1);
        }

        for (Map.Entry<Integer, Integer> e : freq.entrySet()) {
            if (e.getValue() > nums.length / 2)
                majority = e.getKey();
        }
        return majority;
    }

    //https://leetcode.com/problems/happy-number/submissions/
    public boolean isHappy(int n) {
        int result = 0;
        int iteration = 0;
        while (result != 1 && iteration <= 100) {
            result = 0;
            while (n != 0) {
                result += (n % 10) * (n % 10);
                n /= 10;
            }
//            System.out.println(result);
            n = result;
            iteration++;
        }
        if (result != 1)
            return false;
        return true;
    }

    //https://leetcode.com/problems/count-primes/
    private static boolean isPrime(int n) {
        if (n <= 1) return false;
        if (n <= 3) return true;
        if (n % 2 == 0 || n % 3 == 0) return false;
        for (int i = 5; i * i <= n; i += 6) {
            if (n % i == 0 || n % (i + 2) == 0)
                return false;
        }
        return true;
    }

    private static int nextPrime(int n) {
        if (n <= 1) return 2;
        boolean found = false;
        int prime = n;
        while (!found) {
            prime++;

            if (isPrime(prime))
                found = true;
        }
        return prime;
    }

    public int countPrimes(int n) {
        int count = (int) IntStream.iterate(2, Easy::nextPrime).limit(n).filter(i -> i < n).count();
        return count;
    }

    //https://leetcode.com/problems/valid-anagram/submissions/
    public boolean isAnagram(String s, String t) {
        int[] freq1 = new int[256];
        int[] freq2 = new int[256];
        if (s.length() != t.length()) return false;
        if (s.equals(t)) return true;
        for (int i = 0; i < s.length(); i++) {
            freq1[s.charAt(i)]++;
        }
        for (int i = 0; i < t.length(); i++)
            freq2[t.charAt(i)]++;

        for (int i = 0; i < 256; i++) {
            if (freq1[i] != freq2[i])
                return false;
        }
        return true;
    }
}

File: Stefan Alexandru/LeetCode/easy.py
def romanToInt(s):
    number = 0
    dictt = {'I': 1, 'V': 5, 'X': 10, 'L': 50, 'C': 100, 'D': 500, 'M': 1000}
    ok = True
    for i in range(len(s)):
        if not ok:
            ok = True
            continue
        if i != len(s) - 1:
            if dictt[s[i]] < dictt[s[i + 1]]:
                number += dictt[s[i + 1]] - dictt[s[i]]
                ok = False
            else:
                number += dictt[s[i]]
        else:
            if dictt[s[i]] <= dictt[s[i - 1]]:
                number += dictt[s[i]]
    return number


# print(romanToInt('III'))


# https://leetcode.com/problems/valid-parentheses/
def isValid(s: str) -> bool:
    dic = {'(': 0, '{': 0, '[': 0}
    oppose = {')': '(', '}': '{', ']': '['}
    last_open = []
    for paranthesis in s:
        if paranthesis not in dic:
            if dic[oppose[paranthesis]] == 0 or last_open[-1] != oppose[paranthesis]:
                return False
            del last_open[-1]
            dic[oppose[paranthesis]] -= 1
        else:
            dic[paranthesis] += 1
            last_open.append(paranthesis)
    return True if not sum([dic[key] for key in dic]) else False


print(isValid("([])"))


# https://leetcode.com/problems/min-stack/submissions/
class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.vect = []

    def push(self, x: int) -> None:
        self.vect.append(x)

    def pop(self) -> None:
        del self.vect[-1]

    def top(self) -> int:
        return self.vect[-1]

    def getMin(self) -> int:
        return min(self.vect)


# https://leetcode.com/problems/merge-two-sorted-lists/
# https://leetcode.com/problems/linked-list-cycle/submissions/
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

    def mergeTwoLists(self, l1, l2):
        final = ListNode(0)
        copy = final
        while l1 and l2:
            if l1.val > l2.val:
                final.next = ListNode(l2.val)
                l2 = l2.next
            else:
                final.next = ListNode(l1.val)
                l1 = l1.next
            final = final.next
        if l1:
            final.next = l1
        else:
            final.next = l2
        return copy.next

    def hasCycle(self, head):
        first, second = head, head
        while first and first.next:
            first, second = first.next.next, second.next
            if first is second:
                return True
        return False


# https://leetcode.com/problems/majority-element/submissions/
def majorityElement(nums):
    summ = len(nums) / 2
    for elem in set(nums):
        if nums.count(elem) > summ:
            return elem


# https://leetcode.com/problems/happy-number/submissions/
def isHappy(n):
    nrs = []
    while n not in nrs:
        nrs.append(n)
        n_str = sum([int(digit) ** 2 for digit in str(n)])
        if n_str == 1:
            return True
        n = ''.join(str(n_str))
    return False


# https://leetcode.com/problems/valid-anagram/submissions/
def isAnagram(self, s, t):
    """
    :type s: str
    :type t: str
    :rtype: bool
    """
    letters = {}
    for letter in s:
        if letter not in letters:
            letters[letter] = 1
        else:
            letters[letter] += 1

    for letter in t:
        if letter not in letters:
            return False
        else:
            letters[letter] -= 1

    for letter in letters:
        if letters[letter] != 0:
            return False

    return True


# https://leetcode.com/problems/merge-sorted-array/submissions/
def merge(nums1, m, nums2, n):
    """
    Do not return anything, modify nums1 in-place instead.
    """
    last, i, j = m + n - 1, m - 1, n - 1

    while i >= 0 and j >= 0:
        if nums1[i] > nums2[j]:
            nums1[last] = nums1[i]
            last, i = last - 1, i - 1
        else:
            nums1[last] = nums2[j]
            last, j = last - 1, j - 1

    while j >= 0:
        nums1[last] = nums2[j]
        last, j = last - 1, j - 1


# https://leetcode.com/problems/rotate-array/submissions/
def rotate(nums, k):
    """
    Do not return anything, modify nums in-place instead.
    """
    a = [0] * len(nums)
    for i in range(0, len(nums)):
        a[(i + k) % len(nums)] = nums[i]
    for i in range(0, len(nums)):
        nums[i] = a[i]


# https://leetcode.com/problems/convert-sorted-array-to-binary-search-tree/
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

    def sortedArrayToBST(self, num):
        return self.sortedArrayToBSTRecu(num, 0, len(num))

    def sortedArrayToBSTRecu(self, num, start, end):
        if start == end:
            return None
        mid = start + (end - start) / 2
        node = TreeNode(num[mid])
        node.left = self.sortedArrayToBSTRecu(num, start, mid)
        node.right = self.sortedArrayToBSTRecu(num, mid + 1, end)
        return node


# https://leetcode.com/problems/count-primes/
from math import sqrt


def countPrimes(n):
    """
    :type n: int
    :rtype: int
    """
    if n <= 2:
        return 0

    is_prime = [True] * n
    sqr = sqrt(n - 1)

    num = 0
    for i in xrange(2, n):
        if is_prime[i]:
            num += 1
            for j in xrange(i + i, n, i):
                is_prime[j] = False

    return num
